modules/mtl/arrow-mtl/src/main/kotlin/arrow/mtl/extensions/accumt.kt

```diff
@@ -0,0 +1,193 @@
+    AccumT.just(MS(), MF(), a)
+
+  override fun <A, B> Kind<AccumTPartialOf<S, F>, A>.flatMap(f: (A) -> Kind<AccumTPartialOf<S, F>, B>): Kind<AccumTPartialOf<S, F>, B> =
+    this.fix().flatMap(MS(), MF(), f)
+
+  override fun <A, B> tailRecM(a: A, f: (A) -> Kind<AccumTPartialOf<S, F>, Either<A, B>>): Kind<AccumTPartialOf<S, F>, B> =
+    AccumT.tailRecM(MF(), a, f)
+
+  override fun <A, B> Kind<AccumTPartialOf<S, F>, A>.ap(ff: Kind<AccumTPartialOf<S, F>, (A) -> B>): Kind<AccumTPartialOf<S, F>, B> =
+    fix().ap(MS(), MF(), ff)
+}
+
+@extension
+interface AccumtTMonadTrans<S> : MonadTrans<Kind<ForAccumT, S>> {
+
+  fun MS(): Monoid<S>
+
+  override fun <G, A> Kind<G, A>.liftT(MG: Monad<G>): Kind2<Kind<ForAccumT, S>, G, A> =
+    AccumT(MG) { _: S ->
+      MG.run {
+        flatMap { a ->
+          MG.just(MS().empty() toT a)
+        }
+      }
+    }
+}
+
+@extension
+interface AccumTAlternative<S, F> : Alternative<AccumTPartialOf<S, F>>, AccumTApplicative<S, F> {
+
+  fun AF(): Alternative<F>
+  override fun MF(): Monad<F>
+  override fun MS(): Monoid<S>
+
+  override fun <A> Kind<AccumTPartialOf<S, F>, A>.orElse(b: Kind<AccumTPartialOf<S, F>, A>): Kind<AccumTPartialOf<S, F>, A> =
+    (this.fix() to b.fix()).let { (ls, rs) ->
+      AccumT(AF()) { s: S ->
+        AF().run {
+          ls.runAccumT(MF(), s).orElse(rs.runAccumT(MF(), s))
+        }
+      }
+    }
+
+  override fun <A> empty(): Kind<AccumTPartialOf<S, F>, A> =
+    AccumT.liftF(AF(), AF().empty())
+}
+
+@extension
+interface AccumTApplicativeError<S, F, E> : ApplicativeError<AccumTPartialOf<S, F>, E>, AccumTApplicative<S, F> {
+  fun ME(): MonadError<F, E>
+
+  override fun MS(): Monoid<S>
+  override fun MF(): Monad<F> = ME()
+
+  override fun <A> raiseError(e: E): Kind<AccumTPartialOf<S, F>, A> =
+    AccumT.liftF(MF(), ME().raiseError(e))
+
+  override fun <A> Kind<AccumTPartialOf<S, F>, A>.handleErrorWith(f: (E) -> Kind<AccumTPartialOf<S, F>, A>): Kind<AccumTPartialOf<S, F>, A> =
+    this.fix().let { accumT ->
+      AccumT(MF()) { s: S ->
+        ME().run {
+          accumT.runAccumT(MF(), s).handleErrorWith { e ->
+            f(e).fix().runAccumT(MF(), s)
+          }
+        }
+      }
+    }
+}
+
+@extension
+interface AccumTMonadError<S, F, E> : MonadError<AccumTPartialOf<S, F>, E>, AccumTApplicativeError<S, F, E>, AccumTMonad<S, F> {
+  override fun MS(): Monoid<S>
+  override fun ME(): MonadError<F, E>
+  override fun MF(): Monad<F> = ME()
+}
+
+@extension
+interface AccumTMonadState<S, F> : MonadState<AccumTPartialOf<S, F>, S>, AccumTMonad<S, F> {
+  fun MSF(): MonadState<F, S>
+  override fun MF(): Monad<F> = MSF()
+  override fun MS(): Monoid<S>
+
+  override fun get(): Kind<AccumTPartialOf<S, F>, S> =
+    AccumT.liftF(MSF(), MSF().get())
+
+  override fun set(s: S): Kind<AccumTPartialOf<S, F>, Unit> =
+    AccumT.liftF(MSF(), MSF().set(s))
+}
+
+@extension
+interface AccumTMonadReader<S, F> : MonadReader<AccumTPartialOf<S, F>, S>, AccumTMonad<S, F> {
```

Comment:
Same as above, the reader context is not the accumulating value of the `AccumT` it is just some reader value that gets lifted through